### PR TITLE
wait for node to report address before removing taint

### PR DESCRIPTION
**Pull request by huwcbjones**
- **address/assigner: return assigned address**
- **cmd/assignAddress: return assignedAddress**
- **cmd/run: wait for address to be reported before removing taint**

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -130,7 +131,64 @@ func assignAddress(c context.Context, log *logrus.Entry, client kubernetes.Inter
...
+	}
+	return "", errors.New("reached maximum number of retries")
+}
+
+func waitForAddressToBeReported(c context.Context, log *logrus.Entry, explorer nd.Explorer, node *types.Node, assignedAddress string, cfg *config.Config) error {
+	ctx, cancel := context.WithCancel(c)
+	defer cancel()
+
+	// ticker for retry interval
+	ticker := time.NewTicker(cfg.RetryInterval)
```
Comment (w-miller) [inline cmd/main.go:145]: Worth making this separately configurable to the main assignment retry interval? That way we can poll more frequently and be more responsive at this point once we've instructed to perform the assignment.

File: cmd/main.go [hunk 2]
```diff
@@ -130,7 +131,64 @@ func assignAddress(c context.Context, log *logrus.Entry, client kubernetes.Inter
...
+						"retry-attempts": cfg.RetryAttempts,
+					}).Info("Node is reporting assigned address")
+					return nil
+				}
+			}
+			log.WithError(err).WithFields(logrus.Fields{
+				"node":     node.Name,
+				"instance": node.Instance,
+				"address":  assignedAddress,
+			}).Error("Node is not yet reporting the assigned address")
```
Comment (w-miller) [inline cmd/main.go]: Should this be an info/warning rather than an error, since we expect propagation to take some time?

File: cmd/main_test.go [hunk 3]
```diff
@@ -168,8 +174,201 @@ func Test_assignAddress(t *testing.T) {
...
+					mock := nodeMocks.NewExplorer(t)
+					mock.EXPECT().GetNode(tmock.Anything, "test-node").Return(&types.Node{
+						Name:        "test-node",
+						Instance:    "test-instance",
+						Region:      "test-region",
+						Zone:        "test-zone",
+						ExternalIPs: []net.IP{net.IPv4(9, 9, 9, 9)},
+					}, nil).Times(4)
+					return mock
+				},
```
Comment (w-miller) [inline cmd/main_test.go:293]: Worth adding to return the correct IP after the retries, to verify we still get an error because we gave up too early?

File: internal/address/gcp.go [hunk 4]
```diff
@@ -214,20 +214,23 @@ func (a *gcpAssigner) CheckAddressAssigned(region, addressName string) (bool, er
...
 }
 
-func (a *gcpAssigner) Assign(ctx context.Context, instanceID, zone string, filter []string, orderBy string) error {
+func (a *gcpAssigner) Assign(ctx context.Context, instanceID, zone string, filter []string, orderBy string) (string, error) {
 	// check if instance already has a public static IP address assigned
-	instance, err := a.checkStaticIPAssigned(zone, instanceID)
+	instance, address, err := a.checkStaticIPAssigned(zone, instanceID)
 	if err != nil {
-		return errors.Wrapf(err, "check if static public IP is already assigned to instance %s", instanceID)
+		if errors.Is(err, ErrStaticIPAlreadyAssigned) {
```
Comment (w-miller) [inline internal/address/gcp.go:221]: Is this a change in behaviour? If so, should it go in a separate PR?

**Review by w-miller — COMMENTED** [review]
Some minor comments but overall looks reasonable to me!

File: cmd/main.go (hunk 1 above)
Comment (huwcbjones) [reply cmd/main.go:145]: Now that the upstream maintainer has got back to me, I'll leave as-is and see what they say.
For our use-case we can always tweak the Retry interval and counts.

File: internal/address/gcp.go (hunk 4 above)
Comment (huwcbjones) [reply internal/address/gcp.go:221]: Sort of, the `ErrStaticIPAlreadyAssigned` error was being caught by the caller anyway, but we need a way to return the assigned static address so we can wait for it to be reported later.

**Comment by huwcbjones** [comment]
Superseded by doitintl/kubeip#158
